money-core/src/main/scala/com/comcast/money/core/ConfigurableTypeFactory.scala

```diff
@@ -0,0 +1,124 @@
+      .map(_(config))
+      .orElse(defaultValue)
+
+  protected def findConfigValue(config: Config, key: String): Option[String] =
+    Option(config)
+      .filter(_.hasPath(key))
+      .map(_.getString(key))
+      .filterNot(_.isEmpty)
+
+  protected def findKnownType(config: Config): Option[Config => T] =
+    findConfigValue(config, TYPE_KEY)
+      .flatMap(findKnownType)
+
+  protected def findKnownType(typeName: String): Option[Config => T] =
+    knownTypes.lift(typeName)
+      .orElse({
+        logger.warn(s"Could not find ${tag.runtimeClass.getSimpleName} type: '$typeName'")
+        None
+      })
+
+  protected def createInstance(config: Config): Option[Config => T] =
+    findConfigValue(config, CLASS_KEY)
+      .flatMap(findConfigurableTypeFactory)
+
+  protected def findConfigurableTypeFactory(className: String): Option[Config => T] =
+    findClass(className)
+      .filter(tag.runtimeClass.isAssignableFrom)
+      .flatMap({
+        classInstance =>
+          findFactoryMethod(classInstance)
+            .orElse(findConfigConstructor(classInstance))
+            .orElse(findDefaultConstructor(classInstance))
+      })
+      .orElse(findModule(className)
+        .flatMap({
+          moduleInstance => findStaticInstance(moduleInstance)
+        }))
+      .orElse({
+        logger.warn(s"Could not create ${tag.runtimeClass.getSimpleName} class: '$className'")
+        None
+      })
+
+  private def findClass(className: String): Option[Class[_]] =
+    Try(Class.forName(className)).toOption
+
+  private def findModule(moduleName: String): Option[Class[_]] =
+    findClass(moduleName + MODULE_SUFFIX)
+
+  private def findFactoryMethod(cls: Class[_]): Option[Config => T] =
+    cls.getMethods.find {
+      method =>
+        Modifier.isStatic(method.getModifiers) &&
+          method.getParameterCount == 1 &&
+          method.getParameterTypes()(0) == classOf[Config] &&
+          method.getReturnType == cls &&
+          method.getName == APPLY_NAME
+    } map { method => config: Config => method.invoke(null, config).asInstanceOf[T] }
+
+  private def findConfigConstructor(cls: Class[_]): Option[Config => T] =
+    cls.getConstructors.find {
+      constructor =>
+        constructor.getParameterCount == 1 &&
+          constructor.getParameterTypes()(0) == classOf[Config]
+    } map { constructor => config: Config => constructor.newInstance(config).asInstanceOf[T] }
+
+  private def findDefaultConstructor(cls: Class[_]): Option[Config => T] =
+    cls.getConstructors.find {
+      _.getParameterCount == 0
+    } map { constructor => _: Config => constructor.newInstance().asInstanceOf[T] }
+
+  private def findStaticInstance(cls: Class[_]): Option[Config => T] =
+    cls.getFields.find {
+      field =>
+        Modifier.isStatic(field.getModifiers) &&
+          field.getType == cls &&
+          field.getName == MODULE_FIELD_NAME
+    } map { field => _: Config => field.get(null).asInstanceOf[T] }
+}
```

Comment:
How much code does this save us? I feel like the abstraction itself is deeper than what is being abstracted. Or at least that is my first impression. would you agree with that at all?

Reply:
If we want to go this route we might be better served with a macro. but you would need to bear in mind that that api is being fully rewritten for scala 3.

Reply:
It's true, it's a bit more code than what it replaced, but it does two things.

1.  It replaces the separate uses of reflection to manually create instances of a class by name used in the different factory objects so that they all work the same way.
2. It eliminates the separate configuration step (and configuration traits) used to provide configuration to the plugin instance.  Now configuration has to happen as a part of initialization making it impossible to end up with a plugin instance that isn't configured.  This enables plugins to be properly immutable.